submissions/views/xqueue.py

```diff
@@ -0,0 +1,296 @@
+                self.compose_reply(False, "Submission already in process"),
+                status=status.HTTP_409_CONFLICT
+            )
+
+        if external_grader:
+            try:
+                if external_grader.status != "pulled":
+                    external_grader.update_status("pulled")
+
+                submission_data = {
+                    "grader_payload": json.dumps({"grader": external_grader.grader_file_name}),
+                    "student_info": json.dumps({
+                        "anonymous_student_id": str(external_grader.submission.student_item.student_id),
+                        "submission_time": str(int(external_grader.created_at.timestamp())),
+                        "random_seed": 1
+                    }),
+                    "student_response": external_grader.submission.answer
+                }
+
+                payload = {
+                    'xqueue_header': json.dumps({
+                        'submission_id': external_grader.submission.id,
+                        'submission_key': external_grader.pullkey
+                    }),
+                    'xqueue_body': json.dumps(submission_data),
+                    'xqueue_files': json.dumps(get_files_for_grader(external_grader))
+                }
+
+                return Response(
+                    self.compose_reply(True, content=json.dumps(payload)),
+                    status=status.HTTP_200_OK
+                )
+
+            except ValueError as e:
+                return Response(
+                    self.compose_reply(False, f"Error processing submission: {str(e)}"),
+                    status=status.HTTP_400_BAD_REQUEST
+                )
+
+        return Response(
+            self.compose_reply(False, f"Queue '{queue_name}' is empty"),
+            status=status.HTTP_404_NOT_FOUND
+        )
+
+    @action(detail=False, methods=['post'], url_name='put_result')
+    @transaction.atomic
+    def put_result(self, request):
+        """
+        Endpoint for graders to post their results and update submission scores.
+        """
+        (reply_valid, submission_id, submission_key, score_msg, points_earned) = (
+            self.validate_grader_reply(request.data))
+
+        if not reply_valid:
+            log.error("Invalid reply from pull-grader: request.data: %s",
+                      request.data)
+            return Response(
+                self.compose_reply(False, 'Incorrect reply format'),
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            external_grader = ExternalGraderDetail.objects.select_for_update(
+                                                                        nowait=True).get(submission__id=submission_id)
+        except ExternalGraderDetail.DoesNotExist:
+            log.error(
+                "Grader submission_id refers to nonexistent entry in Submission DB: "
+                "grader: %s, submission_key: %s, score_msg: %s",
```

Comment:
This log entry says "grader" but the field being written is "submission_id". Also, please keep the old terminology of "grader_reply" rather than "score_msg", since it contains more than just the score, and it's not just a text message.